### PR TITLE
feat(libflux/core/parser): update mod.rs with parse_record 

### PR DESCRIPTION
**Pull request by ChitlangeSahas**
### Done checklist
- [X] docs/SPEC.md updated
- [X] Test cases written

### CODE REVIEW

File: libflux/src/core/parser/mod.rs [hunk 1]
```diff
@@ -577,6 +577,57 @@ impl Parser {
             MonoType::Invalid
         }
     }
+
+    #[cfg(test)]
+    fn parse_record(&mut self) -> Record {
+        //( "{" [Properties] "}" ) | ( "{" Tvar "with" Properties "}" )
+        let _ = self.expect(TOK_LBRACE);
```
Comment (jlapacik) [inline libflux/src/core/parser/mod.rs:584]: I think you'll need this token for constructing the base node.

File: libflux/src/core/parser/mod.rs [hunk 2]
```diff
@@ -577,6 +577,57 @@ impl Parser {
...
+        // can be (Properties) or (Tvar "with" Properties)
+        let t = self.peek();
+        if t.lit.to_uppercase() != (t.lit) {
+            let properties = self.parse_properties();
+            let record = Record {
+                base: self.base_node_from_token(&t),
```
Comment (jlapacik) [inline libflux/src/core/parser/mod.rs:590]: I think you'll need to use `base_node_from_tokens` here and pass in the tokens for the right and left braces.

File: libflux/src/core/parser/mod.rs [hunk 3]
```diff
@@ -577,6 +577,57 @@ impl Parser {
...
         }
     }
+
+    #[cfg(test)]
+    fn parse_record(&mut self) -> Record {
+        //( "{" [Properties] "}" ) | ( "{" Tvar "with" Properties "}" )
+        let _ = self.expect(TOK_LBRACE);
+        // can be (Properties) or (Tvar "with" Properties)
+        let t = self.peek();
+        if t.lit.to_uppercase() != (t.lit) {
```
Comment (jlapacik) [inline libflux/src/core/parser/mod.rs:587]: I don't think this will parse `{A:int}` correctly.

File: libflux/src/core/ast/mod.rs [hunk 4]
```diff
@@ -531,7 +531,24 @@ pub struct TypeExpression {
...
     pub monotype: MonoType,
 }
-
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Record {
+    #[serde(skip_serializing_if = "BaseNode::is_empty")]
+    #[serde(default)]
+    #[serde(flatten)]
+    pub base: BaseNode,
+    pub tvar: Option<MonoType>,
```
Comment (jlapacik) [inline libflux/src/core/ast/mod.rs:540]: If we're checking to make sure this is a type variable instead of just an identifier, then this should probably be

    pub tvar: Option<TvarType>,

File: libflux/src/core/parser/mod.rs [hunk 5]
```diff
@@ -577,6 +577,57 @@ impl Parser {
...
+        }
+        return Some(properties);
+    }
+    #[cfg(test)]
+    fn parse_property(&mut self) -> PropertyType {
+        let identifier = self.parse_identifier(); // identifier
+        self.expect(TOK_COLON); // :
+        let monotype = self.parse_monotype();
+        PropertyType {
+            base: base_from_monotype(&monotype),
```
Comment (jlapacik) [inline libflux/src/core/parser/mod.rs:626]: The base node here should extend from the identifier to the monotype.

File: libflux/src/core/parser/mod.rs [hunk 6]
```diff
@@ -577,6 +577,76 @@ impl Parser {
...
     }
+
+    // Record = "{" [ Identifier (Suffix1 | Suffix2) ] "}"
+    // Suffix1 = ":" MonoType { "," Property }
+    // Suffix2 = "with" [Properties]
+    #[cfg(test)]
+    fn parse_record(&mut self) -> Record {
+        let start = self.open(TOK_LBRACE, TOK_RBRACE);
+        let mut properties: Option<Vec<PropertyType>> = None;
+        let property: PropertyType;
```
Comment (jlapacik) [inline libflux/src/core/parser/mod.rs:588]: No need for this forward declaration.

File: libflux/src/core/parser/mod.rs [hunk 7]
```diff
@@ -577,6 +577,76 @@ impl Parser {
...
+
+        let mut id: Option<Identifier> = None;
+        let _id;
```
Comment (jlapacik) [inline libflux/src/core/parser/mod.rs:591]: No need for this forward declaration either.

File: libflux/src/core/parser/mod.rs [hunk 8]
```diff
@@ -577,6 +577,76 @@ impl Parser {
...
+        let t = self.peek();
+        if t.tok == TOK_IDENT {
+            // Indentifier
+            _id = self.parse_identifier();
+            // suffix one needs attention
+            let t2 = self.peek();
+            if t2.tok == TOK_COLON {
+                let mut ps = Vec::<PropertyType>::new();
+                self.consume(); // :
```
Comment (jlapacik) [inline libflux/src/core/parser/mod.rs:600]: Can you use `self.expect` here for edge cases?

File: libflux/src/core/parser/mod.rs [hunk 9]
```diff
@@ -577,6 +577,76 @@ impl Parser {
...
+        }
+        return Some(properties);
+    }
+    #[cfg(test)]
+    fn parse_property(&mut self) -> PropertyType {
+        let identifier = self.parse_identifier(); // identifier
+        self.expect(TOK_COLON); // :
+        let monotype = self.parse_monotype();
+        PropertyType {
+            base: base_from_monotype(&monotype),
```
Comment (jlapacik) [inline libflux/src/core/parser/mod.rs:645]: Can you change this to `self.base_node_from_others(&identifier.base, &monotype.base)`?

File: libflux/src/core/parser/tests.rs [hunk 10]
```diff
@@ -1466,6 +1466,172 @@ fn test_parse_type_expression_regexp() {
...
+        Record {
+            base: BaseNode {
+                location: loc.get(1, 1, 1, 16),
+                ..BaseNode::default()
+            },
+            tvar: None,
+            properties: Some(vec![
+                PropertyType {
+                    base: BaseNode {
+                        location: loc.get(1, 4, 1, 7),
```
Comment (jlapacik) [inline libflux/src/core/parser/tests.rs:1485]: Similar to my previous comment. The source location for this property should correspond to the entire property `a:int`, but right now it just refers to the location of the monotype `int`.

File: libflux/src/core/parser/mod.rs (hunk 6 above)
Comment (ChitlangeSahas) [reply libflux/src/core/parser/mod.rs:588]: Fixed!

File: libflux/src/core/parser/mod.rs (hunk 7 above)
Comment (ChitlangeSahas) [reply libflux/src/core/parser/mod.rs:591]: I could be wrong about this one, how would I reference it in the else block when it goes out of scope?

File: libflux/src/core/parser/mod.rs (hunk 8 above)
Comment (ChitlangeSahas) [reply libflux/src/core/parser/mod.rs:600]: ✅

File: libflux/src/core/parser/mod.rs (hunk 9 above)
Comment (ChitlangeSahas) [reply libflux/src/core/parser/mod.rs:645]: 👍

File: libflux/src/core/parser/mod.rs (hunk 7 above)
Comment (jlapacik) [reply libflux/src/core/parser/mod.rs:591]: You assign it in the outermost if block, so it's still visible from inside the inner if/else blocks.

Comment (ChitlangeSahas) [reply libflux/src/core/parser/mod.rs]: oh very true.

File: libflux/src/core/parser/mod.rs (hunk 8 above)
Comment (jlapacik) [reply libflux/src/core/parser/mod.rs]: Oops, you already check for a `:` before consuming. My bad.
